toolchain/parser/parser_handle_paren_condition.cpp

```diff
@@ -12,11 +12,23 @@ static auto ParserHandleParenCondition(ParserContext& context,
                                        ParserState finish_state) -> void {
   auto state = context.PopState();
 
-  context.ConsumeAndAddOpenParen(state.token, start_kind);
-
+  std::optional<TokenizedBuffer::Token> open_paren =
+      context.ConsumeAndAddOpenParen(state.token, start_kind);
+  if (open_paren) {
+    state.token = *open_paren;
+  }
   state.state = finish_state;
   context.PushState(state);
-  context.PushState(ParserState::Expression);
+
+  if (!open_paren && context.PositionIs(TokenKind::OpenCurlyBrace)) {
+    // For an open curly, assume the condition was completely omitted.
+    // Expression parsing would treat the { as a struct, but instead assume it's
+    // a code block and just emit an invalid parse.
+    context.AddLeafNode(ParseNodeKind::InvalidParse, *context.position(),
+                        /*has_error=*/true);
+  } else {
+    context.PushState(ParserState::Expression);
```

Comment:
Do the comments on `ParenCondition` in `parser_state.def` need to be updated now that this push isn't unconditional?

Reply:
Thanks for the reminder, done.